riscv/triggers.cc

```diff
@@ -0,0 +1,206 @@
+  switch (match_value) {
+    case MATCH_EQUAL:
+    case MATCH_NAPOT:
+    case MATCH_GE:
+    case MATCH_LT:
+    case MATCH_MASK_LOW:
+    case MATCH_MASK_HIGH:
+      match = (triggers::mcontrol_t::match_t) match_value;
+      break;
+    default:
+      match = MATCH_EQUAL;
+      break;
+  }
+  m = get_field(val, MCONTROL_M);
+  h = get_field(val, MCONTROL_H);
+  s = get_field(val, MCONTROL_S);
+  u = get_field(val, MCONTROL_U);
+  execute_bit = get_field(val, MCONTROL_EXECUTE);
+  store_bit = get_field(val, MCONTROL_STORE);
+  load_bit = get_field(val, MCONTROL_LOAD);
+  // Assume we're here because of csrw.
+  if (execute_bit)
+    timing = 0;
+  return true;
+}
+
+reg_t mcontrol_t::tdata2_read(const processor_t * const proc) const noexcept {
+  return tdata2;
+}
+
+bool mcontrol_t::tdata2_write(processor_t * const proc, const reg_t val) noexcept {
+  if (dmode && !proc->get_state()->debug_mode) {
+    return false;
+  }
+  tdata2 = val;
+  return true;
+}
+
+bool mcontrol_t::simple_match(unsigned xlen, reg_t value) const {
+  switch (match) {
+    case triggers::mcontrol_t::MATCH_EQUAL:
+      return value == tdata2;
+    case triggers::mcontrol_t::MATCH_NAPOT:
+      {
+        reg_t mask = ~((1 << (cto(tdata2)+1)) - 1);
+        return (value & mask) == (tdata2 & mask);
+      }
+    case triggers::mcontrol_t::MATCH_GE:
+      return value >= tdata2;
+    case triggers::mcontrol_t::MATCH_LT:
+      return value < tdata2;
+    case triggers::mcontrol_t::MATCH_MASK_LOW:
+      {
+        reg_t mask = tdata2 >> (xlen/2);
+        return (value & mask) == (tdata2 & mask);
+      }
+    case triggers::mcontrol_t::MATCH_MASK_HIGH:
+      {
+        reg_t mask = tdata2 >> (xlen/2);
+        return ((value >> (xlen/2)) & mask) == (tdata2 & mask);
+      }
+  }
+  assert(0);
+}
+
+match_result_t mcontrol_t::memory_access_match(processor_t * const proc, operation_t operation, reg_t address, reg_t data) {
+  state_t * const state = proc->get_state();
+  if ((operation == triggers::OPERATION_EXECUTE && !execute_bit) ||
+      (operation == triggers::OPERATION_STORE && !store_bit) ||
+      (operation == triggers::OPERATION_LOAD && !load_bit) ||
+      (state->prv == PRV_M && !m) ||
+      (state->prv == PRV_S && !s) ||
+      (state->prv == PRV_U && !u)) {
+    return MATCH_NONE;
+  }
+
+  reg_t value;
+  if (select) {
+    value = data;
+  } else {
+    value = address;
+  }
+
+  // We need this because in 32-bit mode sometimes the PC bits get sign
+  // extended.
+  auto xlen = proc->get_xlen();
+  if (xlen == 32) {
+    value &= 0xffffffff;
+  }
+
+  if (simple_match(xlen, value)) {
+    if (timing)
+      return MATCH_FIRE_AFTER;
+    else
+      return MATCH_FIRE_BEFORE;
+  }
+  return MATCH_NONE;
+}
+
+module_t::module_t(unsigned count) : triggers(count) {
+  for (unsigned i = 0; i < count; i++) {
+    triggers[i] = new mcontrol_t();
+  }
+}
+
+module_t::~module_t() {
+  for (unsigned i = 0; i < triggers.size(); i++) {
```

Comment:
Any reason we couldn't use a modern loop e.g. `for (auto trigger : triggers)`?

Reply:
Done.